nixos/modules/services/misc/matrix-synapse.nix

```diff
@@ -4,120 +4,19 @@ with lib;
 
 let
   cfg = config.services.matrix-synapse;
-  opt = options.services.matrix-synapse;
-  pg = config.services.postgresql;
-  usePostgresql = cfg.database_type == "psycopg2";
-  logConfigFile = pkgs.writeText "log_config.yaml" cfg.logConfig;
-  mkResource = r: ''{names: ${builtins.toJSON r.names}, compress: ${boolToString r.compress}}'';
-  mkListener = l: ''{port: ${toString l.port}, bind_address: "${l.bind_address}", type: ${l.type}, tls: ${boolToString l.tls}, x_forwarded: ${boolToString l.x_forwarded}, resources: [${concatStringsSep "," (map mkResource l.resources)}]}'';
+  format = pkgs.formats.yaml {};
+
+  # remove null values from the final configuration
+  finalSettings = lib.converge (lib.filterAttrsRecursive (_: v: ! elem v [ null ])) cfg.settings;
```

Comment:
I think this is problematic due to things like:

```
# How long to keep redacted events in unredacted form in the database. After
# this period redacted events get replaced with their redacted form in the DB.
#
# Defaults to `7d`. Set to `null` to disable.
#
#redaction_retention_period: 28d
```

it sucks that there really is no better way of getting this functionality

or is this solved by setting options to empty string?

Reply:
You could get away with setting `"null"` as a string, as opposed to `null` the null type. That's not really pretty, but I don't have any great ideas right now how to improve on that.